묵찌빠 프로젝트 [STEP 2] 노움, 에피

Gnoam-R opened the pull request:
@minsson 안녕하세요. 에피, 노움입니다. @hyeffie @rohhyungwoo 

이번 스텝에서는 지난 리뷰를 참고해 게임의 정상적인 진행을 코드로서 보장하고, 불필요한 객체을 생성하지 않고, 타입으로 객체의 역할을 보장 및 제한하는 것을 중점으로 작업을 진행했습니다.

추가 요구사항을 반영하면서도, 코드 재사용성을 높이고, 코드 변경을 최소화하는 방식으로 설계하려고 노력했습니다. 이 부분을 중점적으로 리뷰해주시면 도움이 많이 될 것 같습니다!

## 설계 설명

### HandGameApp

- 앱은 플레이어를 입력 받아 가위바위보 게임과 묵찌빠 게임을 진행합니다.
- 생성 시 에러를 표시할 수 있는 객체를 전달받아 게임 진행 중 발생한 에러(예외) 관련 메시지를 출력합니다.

### RPSGame, MJBGame

- 각 게임 구조체는 게임을 진행할 두 명의 플레이어로 생성됩니다.
- 게임은 루프를 돌며 각 게임의 Part 게임을 진행합니다.
- Game은 두 플레이어(RPSPlayable)로부터 Gesture 를 생성해 Part 게임을 진행하고, Part 게임의 결과를 각 플레이어(RPSResultDisplayable)가 출력하도록 합니다.

### RPSPart, MJBPart

- 각 게임의 Part 게임은 플레이어 두 명의 Gesture을 가지고 생성됩니다.
- **Gesture**는 플레이어가 제출한 Hand와, Hand를 제출한 플레이어 정보로 구성되어 있습니다.
- Part 는 전달받은 Hand 를 가지고 파트 게임 결과를 판정해, 게임 결과와 게임의 승자 정보를 표현하는 **RPSResult** 타입을 반환합니다.

### UserPlayer

사용자 플레이어는 다음의 기능을 가진 플레이어입니다.

- **RPSPlayable, MJBPlayable**
    
    가위바위보 게임을 위해 Gesture를 제출할 수 있습니다.
    
- **RPSResultDisplayable, MJBResultDisplayable**
    
    가위바위보와 묵찌빠 게임의 결과를 출력할 수 있습니다.
    
    **ComputerPlayer**는 UserPlayer와 다르게 ResultDisplayable을 구현하고 있지 않기 때문에 결과를 출력할 수 없습니다.
    
- **CallablePlayer**
    
    묵찌빠 게임에서 각 플레이어가 현재 Game의 turn을 출력할 때 필요한 이름을 가지고 있습니다.
    

### Console

입출력을 받는 console은 다음의 기능을 가지고 있습니다.

- **InputGettable**
    
    실제 콘솔로부터 사용자 입력을 받아 가공해서 리턴합니다.
    
- **PromptDisplayable**
    
    콘솔에 전달된 프롬프트 문자열을 출력할 수 있습니다.
    
- **OuputDisplayble**
    
    콘솔에 결과 문자열을 출력할 수 있습니다.
    
- **HandGameErrorDisplayble**
    
    콘솔에 전달된 에러의 메시지를 출력할 수 있습니다.
    

## 질문

### Player 가 IO를 가지고 있는 구조, 괜찮을까?

지금은 플레이어가 직접 자신의 인터페이스를 사용해 입력을 받고, 결과를 출력할 수 있도록 의도적으로 IO를 가지고 있도록 설계했는데요.

프로토콜을 사용해 결합도를 낮추고 있기는 하지만, 입출력을 소유하고 있는 구조에 대해서 스스로도 맞는 설계라고 확신하지 못하고 있습니다.

이 문제에 대해서 민쏜은 어떻게 생각하시는지 궁금합니다!

### class로 구현된 플레이어를 sturct로 변경할 수 있을까?

현재 구현에서는 Playable 타입의 플레이어를 동일 비교하는 로직이 필요해서, RPSPlayable과 MJBPlayable을 클래스 전용 프로토콜로 선언하고 있는데요.

```swift
protocol RPSPlayable: AnyObject {
    func makeRPSGesture() throws -> RPSGesture
}
```

지난 리뷰를 참고하다보니 Player 객체들이 클래스로 선언되어야 할 이유가 없어서 구조체로 변경을 시도해보았습니다.

그런데 프로토콜 타입은 Equatable을 직접 구현할 수 없어서 결과적으로 구조체로 코드를 변경하는 데 실패했는데요. 

혹시 플레이어를 동일 비교가 아닌 동등 비교할 수 있다고 볼 수 있을지, 그리고 현재 코드에서 클래스로 선언된 플레이어를 구조체로 변경할 수 있는 방법이 있는지 궁금합니다.

### 플레이어의 생성 시점

App은 가위바위보 게임과 묵찌빠 게임을 진행하기 위해서 게임을 할 수 있는(HandGamePlayable) 두 명의 플레이어와, 게임 진행 중 발생하는 에러의 메시지를 출력하기 위한(HandGameErrorDisplayble) 객체를 생성 시점에 전달받고 있는데요.

```swift
// main.swift
let console = Console()
let duo = HandGameDuo(UserPlayer(io: console, name: "에피"), ComputerPlayer())
HandGameApp(playerDuo: duo, errorDisplay: console).run()
```

그런데 객체와 메소드의 의미를 따져보았을 때, App 을 실행하기 전에 플레이어가 생성되는 것이 맞나? 앱을 실행했을 때 플레이어가 생성되어야 하는 것이 아닐까? 라는 의문이 들었습니다.

앱을 실행하는 시점에 플레이어를 생성하는 로직으로도 바꿔보았는데, 이렇게 되면 앱의 역할이 무엇인가? 이 로직들을 main에 나열하는 것과는 어떻게 다른가? 라는 또 다른 의문이 생기더라고요.

현재 App의 구현으로 보았을 때, 플레이어의 생성 시점은 어느 위치가 적절하다고 보시는지 민쏜의 의견이 궁금합니다.

### 플레이어들이 묵찌빠를 할 수 있는지 확인하는 로직의 위치

지금 App의 run 메소드에서는 다음의 순서로 게임을 진행하고 있습니다.

1. 가위바위보 게임을 진행
2. 가위바위보의 승자가 묵찌빠 게임을 할 수 있는지 확인
3. 가능하다면 묵찌빠 게임 진행

```swift
struct HandGameApp {
    ...
    func run() {
        do {
            let rpsGame = RPSGame(between: playerDuo.leftPlayer, and: playerDuo.rightPlayer)
            let rpsWinner = try rpsGame.start()
            let mjbPlayers = try playerDuo.prepareMJBPlayers(rpsWinner: rpsWinner)
            var mjbGame = MJBGame(
                turn: mjbPlayers.winner,
                other: mjbPlayers.loser
            )
            try mjbGame.start()
        } catch HandGameError.someoneWantsToExit {
            errorDisplay.displayRPSError(HandGameError.someoneWantsToExit)
        } catch { return }
    }
}
```

그런데 이때 HandGameDuo 가 스스로가 묵찌빠 게임을 할 수 있는 지 확인하고, 게임을 할 수 있는 타입(MJBPlayable)으로 타입을 교체하는 것도 어색할 뿐더러 요구사항과 관련 없는 에러를 던지고 처리하고 있다는 생각이 들어 코드를 변경했습니다.

현실의 묵찌빠 게임도 가위바위보를 거쳐서 진행된다는 점에 착안해, MJBGame을 생성하는 시점에 전달된 가위바위보 플레이어들이 묵찌빠를 진행할 수 있는지 판단하도록 로직을 이동시켰습니다.

```swift
// HandGameApp

private func startGame(with leftPlayer: HandGamePlayable, and rightPlayer: HandGamePlayable) {
        do {
            let rpsGame = RPSGame(between: leftPlayer, and: rightPlayer)
            let (rpsWinner, rpsLoser) = try rpsGame.start()
            guard var mjbGame = MJBGame(rpsWinner: rpsWinner, rpsLoser: rpsLoser) else { return }
            try mjbGame.start()
        } catch HandGameError.someoneWantsToExit {
            errorDisplay.displayRPSError(HandGameError.someoneWantsToExit)
        } catch { return }
    }
```

```swift
struct MJBGame {
    ...
    init?(rpsWinner: RPSPlayable,rpsLoser: RPSPlayable) {
        guard let turn = rpsWinner as? MJBPlayable,
              let other = rpsLoser as? MJBPlayable else { return nil }
        self.turn = turn
        self.leftPlayer = self.turn
        self.rightPlayer = other
    }
```

사실 이 로직은 App에서 가위바위보 게임에서 결과가 리턴되고 묵찌빠 게임을 생성하기 전에 해도 무방한 것 같습니다. 민쏜의 의견은 어떠신지 궁금하네요.

hyeffie committed:
feat #1: draft

hyeffie committed:
fix #1: MJBGame play 로직 디버깅

- continue 위치 변경
- printOutput 이름 수정

hyeffie committed:
chore #1: judge 로직 이동

- judge 로직 이동 Hand -> HandGameRound
- isRunning 속성 제거

hyeffie committed:
refactor #1: MJBGame class > struct 로 변경

hyeffie committed:
refactor #1: 에러 처리 로직 수정

- catch 블럭 안에서 switch 문 쓰던 구현에서 특정 에러를 catch해서 처리하고 이외는 다시 에러를 던지도록 구현

hyeffie committed:
refactor #1: RSPApp class > struct 변경 등

- Hand 윈시값 제거
- ConsoleManager final 키워드 추가
- RSPApp class > struct

hyeffie committed:
feat #2: 가위바위보 로직 완성

- 사용자 관점 없도록 설계
- 가위바위보 로직까지 완성
- Playable 타입 지원으로 두 플레이어 모두 입력 받을 수 있도록 구현

hyeffie committed:
refactor #2: PlayerDecision 구현 수정

- PlayerDecision이 연관값으로 Hand 대신 Gesture를 갖도록 수정
- 리터럴 수정

hyeffie committed:
feat #2: 묵찌빠 구현

hyeffie committed:
refactor #2: 앱 레벨 display 속성 타입 변경

- RPSApp의 display 프로퍼티 타입 변경 IO > Printable
- "print" > "display"
- UserPlayer, ComputerPlayer 의 name 접근 레벨 변경

hyeffie committed:
rename #2: 파일 분리 및 renaming

hyeffie committed:
feat #4: MJBResultDisplayable 구현

- MJBResultDisplayable 선언 수정
- UserPlayer 에 MJBResultDisplayable 구현 추가
- CallablePlayer 추가 및 구현
- MJBResult description 속성 추가

hyeffie committed:
chore #4: UserPlayer의 RPSResultDisplayable 구현 정리

- 원래 RPSResultDisplayable 처럼 RPSResult에 description을 구현하려고 시도했으나
- player의 이름이 불필요하기도 하고,
- 로직 상 self와 비교하는 로직이 필요해서 불가능하다는 것을 알게 됨

hyeffie committed:
refactor #4: makeMJBDecision 호출시 MJBPlayable 전달하도록 수정

- MJBPlayable의 makeMJBDecision 호출시 MJBPlayable 전달하도록 파라미터 추가
- ComputerPlayer의 경우 이 파라미터를 사용하지 않아서 좋은 구현인지는 모르겠음
- CallablePlayer 요구사항 수정 getName > name

hyeffie committed:
Merge branch 'step2/4-mjb-result-displayable' into step2/2-viewfree

hyeffie committed:
feat #3: 가위바위보 게임 즉시 종료 구현

- RPSDecision 제거
- 관련 로직 수정

hyeffie committed:
feat #3: 묵찌빠 게임 즉시 종료 구현

- MJBDecision 제거
- 로직 수정
- getRPSGesture 이름 변경

hyeffie committed:
remove #4: MJBGame display 속성 제거

hyeffie committed:
refactor #4: App의 display 속성 타입 변경

- 네이밍 변경
- App의 display 타입 변경 Displayable > HandGameErrorDisplayble

hyeffie committed:
Merge branch 'step2/3-direct-exit' into step2/2-viewfree

hyeffie committed:
remove #3: MJBDecision 제거

hyeffie committed:
chore #3: App 에러 처리 로직 수정

hyeffie committed:
refactor #5: Playable 변환 로직 이동

hyeffie committed:
refactor #5: App 외부에서 플레이어를 생성하는 로직 제거

- HandGameDuo 제거
- 플레이어 생성 로직 이동 main > App

hyeffie committed:
refactor #5: 플레이어 타입 변환 로직 이동

App > MJBGame init

hyeffie committed:
Merge branch 'step2/5-convert-logic' into step2-effie

Gnoam-R committed:
Merge pull request #2 from hyeffie/step2-effie

STEP 2 요구사항 구현

hyeffie committed:
Update README.md

프로젝트 마지막 날...

minsson commented on RockPaperScissors/RockPaperScissors/IO/Console.swift:
```diff
@@ -0,0 +1,35 @@
+
+import Foundation
+
+final class Console: InputGettable {
```
final class를 struct로 바꿔도 그대로 잘 작동하는데, class여야 할 이유가 있을까요?

Gnoam-R replied:
저도 class일 필요가 없다고 생각해요. 클래스를 사용할지 구조체를 사용할지에 대한 의문 그리고 제가 생각한 결론은 객체가 프로퍼티를 가지고 있고 그 프로퍼티가 변동이 될 수 있는 상태값일 때 즉 객체의 상태를 App에서 알아야 할 때 클래스를 사용해야 된다고 생각합니다.

따라서 위 코드에서 class가 아닌 struct가 더 적절하다고 판단됩니다.

hyeffie replied:
말씀하신 것처럼 Console 은 바꿔도 아무 문제 없는 것 같습니다 🥲 반영해두겠습니다!

minsson replied:
넵, 이 문서 한번 참고해보시면 좋을 것 같아요!
[Choosing Between Structures and Classes](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes)

minsson commented on RockPaperScissors/RockPaperScissors/IO/Console.swift:
```diff
@@ -0,0 +1,35 @@
+
+import Foundation
+
+final class Console: InputGettable {
+    func getInput() throws -> String {
+        guard let pureInput = readLine() else { throw HandGameError.invalidInput }
+        let refinedInput = pureInput.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard refinedInput.isEmpty == false else { throw HandGameError.invalidInput }
+        return refinedInput
+    }
+}
```
```suggestion
final class Console { }

extension Console: InputGettable {
    func getInput() throws -> String {
        guard let pureInput = readLine() else { throw HandGameError.invalidInput }
        let refinedInput = pureInput.trimmingCharacters(in: .whitespacesAndNewlines)
        guard refinedInput.isEmpty == false else { throw HandGameError.invalidInput }
        return refinedInput
    }
}
```

minsson replied:
사소한 거지만, 이런 건 아래 코드와 결을 맞춰주는 게 더 깔끔해보여요. 프로토콜의 Requirement를 구현한 거라는 맥락도 줄 수 있고요.

hyeffie replied:
오 저는 항상 저런 식으로 선언부에서부터 구현하는 방식만 사용했었는데 나쁘지 않은 방법인 것 같아요. 반영해보겠습니다!

minsson replied:
이 부분은 취향의 영역이긴 한데, 프로토콜을 채택하고 그 프로토콜의 Requirements를 작성하는 경우, 저는 extension으로 따로 빼는 편이에요. 코드도 더 깔끔해지고, 어떤 프로토콜로 인한 메서드인지도 보기 좋고요.

minsson commented on RockPaperScissors/RockPaperScissors/Decision/MJBGesture.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct MJBGesture {
```
Gesture라는 이름이 좀 어색하게 느껴지네요. UIKit, SwiftUI에서의 TapGesture 등이 연상돼요. 더 명확하게 개선할 수 있을까요?

Gnoam-R replied:
PRS Hand, MJB Hand가 더 적절한거 같아요

hyeffie replied:
Gesture는 Hand랑 플레이어 정보만 가지고 있잖아요?  ‘Part 게임을 위해 각 플레이어가 한 번 제출하는 패’라는 의미를 가진 용어를 사용하고 싶은데 도무지 떠오르지 않더라고요🥲

저도 제스쳐라는 표현이 어색하다고 생각은 해서,, 좀 더 괜찮은 단어를 골라보겠습니다!

minsson commented on RockPaperScissors/RockPaperScissors/Decision/MJBGesture.swift:
```diff
@@ -0,0 +1,13 @@
+//  Created by Effie on 11/28/23.
+//
+
+import Foundation
```
이 코드의 역할은 무엇인가요? 지우면 어떻게 되나요?

Gnoam-R replied:
import Foundation은 기본 데이터 타입, 컬렉션, 사용자 인터페이스 등을 다양한 기능을 사용하는데 필요합니다. 
하지만 위 코드에서 Foundation은 불 필요하다고 생각되네요

hyeffie replied:
정말로 아무 생각 없이 남겨뒀던 것 같아요. 해보니까 모든 파일에서 지워도 문제가 없더라고요…! 반영해두겠습니다.

minsson replied:
넵, 사실 저거 하나 지운다고 성능이 유의미하게 좋아진다든가, 가독성이 엄청나게 좋아진다든가 하는 건 아니지만, 불필요한 코드는 남겨놓지 않는 습관이 중요한 것 같아서 말씀드렸어요!

minsson commented on RockPaperScissors/RockPaperScissors/Game/MJBGame.swift:
```diff
@@ -0,0 +1,61 @@
+        return self.turn === leftPlayer ? rightPlayer : leftPlayer
+    }
+    
+    init?(rpsWinner: RPSPlayable,rpsLoser: RPSPlayable) {
```
종합 코멘트에서도 언급했던 것처럼, 여기서 실패가능한 이니셜라이저를 사용해 묵찌빠 게임 진행 가능 여부를 결정하는 게 적절한지 모르겠어요. 

가독성 측면도 그렇고, 객체 역할 측면에서도 MJBGame은 말그대로 묵찌빠 게임을 진행하는 역할만 하는 게 더 좋을 것 같아요. 따라서, 상위에서 진행 가능 여부를 결정하고, 가능하면 MJBGame을 실행하는 방향이 더 적절하다고 생각해요.

Gnoam-R replied:
말씀 하신 대로 묵찌빠 게임의 진행 가능 여부를 결정하는것은 앱에서 처리하는게 맞다고 생각됩니다.

hyeffie replied:
원래 이 기능을 위해 또 다른 객체를 만들었다가 > App의 역할로 옮겼다가 > MJBGame의 이니셜라이저로 옮기는 여정이 있었는데요. 

이니셜라이저를 옮겼던 의도는 현실에서도 가위바위보를 해야 묵찌빠가 가능하다! 라는 점을 반영하스고자 한 것이었지만. 다시 생각해보면 MJBGame이 그걸 알아야 한다면 두 게임을 타입으로 분리하는 게 어색해지는 것 같다는 생각이 들어요.

App으로 다시 이동시켜보겠습니다!

minsson commented on RockPaperScissors/RockPaperScissors/Game/MJBGame.swift:
```diff
@@ -0,0 +1,61 @@
+    }
+    
+    private func displayResult(_ result: MJBResult) {
+        [leftPlayer, rightPlayer].forEach { player in
```
고차함수도 자연스럽게 사용하시는군요..! 👏👏👏

minsson commented on RockPaperScissors/RockPaperScissors/Game/MJBGame.swift:
```diff
@@ -0,0 +1,61 @@
+        return (turnGesture, otherGesture)
+    }
+    
+    private func displayResult(_ result: MJBResult) {
+        [leftPlayer, rightPlayer].forEach { player in
+            if let displayablePlayer = player as? MJBResultDisplayable {
+                displayablePlayer.display(result: result)
+            }
+        }
+    }
```
Game이 Player에게 '게임 결과를 display해라'라고 메시지를 보내는 게 어색해보여요. 종합코멘트에서 남겼던 것처럼 이런 로직을 Player가 갖고 있는 건 좋지 않은 구조 같아요.

Gnoam-R replied:
앱에 너무 많은 로직이 처리 되는것이 좋지 않다고 생각되어서 이런 구조가 된거 같아요
게임 결과를 출력하는 로직은 App이 가져야 한다고 생각합니다.

hyeffie replied:
이 부분은 Player 가 IO를 가지고 있는 구조, 괜찮을까? 에 대한 재답변으로 갈음할게요!

minsson commented on RockPaperScissors/RockPaperScissors/Game/RPSGame.swift:
```diff
@@ -0,0 +1,48 @@
+    
+    private let rightPlayer: RPSPlayable
+    
+    init(between leftPlayer: RPSPlayable, and rightPlayer: RPSPlayable) { 
```
구조를 떠나, 네이밍이 무척 예쁘다고 생각했어요 👏👏👏

minsson commented on RockPaperScissors/RockPaperScissors/HandGameApp.swift:
```diff
@@ -0,0 +1,38 @@
+            let (rpsWinner, rpsLoser) = try rpsGame.start()
+            guard var mjbGame = MJBGame(rpsWinner: rpsWinner, rpsLoser: rpsLoser) else { return }
+            try mjbGame.start()
+        } catch HandGameError.someoneWantsToExit {
+            errorDisplay.displayRPSError(HandGameError.someoneWantsToExit)
```
이게 왜 Error로 처리되는 건지 잘 이해되지 않아요. 게임종료는 기능 아닐까요?

hyeffie replied:
```swift
enum PlayerDecision {
    case go(gesture: Gesture)
    case stop
}
```

종료 상황을 위해서 중간에 게임을 진행할지 / 종료할지를 표현하는 Decision이라는 타입을 만들어 사용하기도 했지만, 호출 스택이 쌓이는 구조상 리턴이나 throw를 사용하는 것이 깔끔하게 느껴져서 제거하고 구현을 바꿨었던 것 같아요. 로직이 앱쪽으로 이동하는 게 의도한 구조가 아니기도 했고요.

스위프트에서는 Error 타입만 던질 수 있지만, 저는 종료 상황을 예외로 보고 예외를 던지는 로직으로 생각해서 이런 구현을 했었어요.

minsson replied:
저는 게임종료도 게임이 응당 가져야할 하나의 기능이라고 생각해서 '예외'로 처리하는 게 잘 이해되지 않았어요. 예외라는 건 결국 우리가 의도하지 않았거나 예상하지 못한 상황으로 흘러가는 걸 대비하기 위한 건데, 게임 종료는 그 범주에 해당되지 않는다고 생각해요. 

보는 관점에 따라 다를 수 있고, 에피의 입장도 어떤 의도로 말씀하시는 건지 이해가 가기도 해요. 다만 코드를 깔끔하게 만들기 위해서 다른 개발자들이 예상하기 어려운 방향으로 코드를 작성하는 건 조금 조심해야할 수도 있을 것 같아요. 우리는 짧은 코드보다는 읽기 좋은 코드를 써야 하니까요.

minsson commented on RockPaperScissors/RockPaperScissors/HandGameError.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+enum HandGameError: Error {
+    case invalidInput
+    case someoneWantsToExit
+    
+    var description: String {
+        switch self {
+        case .invalidInput: return "잘못된 입력입니다. 다시 시도해주세요."
+        case .someoneWantsToExit: return "게임 종료"
+        }
+    }
+}
```
someoneWantsToExit라는 네이밍은 약간의 위트를 담아서 만드신 걸까요..!? 😅 재미있기는 하지만 직관적으로 읽히지는 않아서 약간의 트레이드 오프가 있었던 것 같아요 😁

또, someoneWantsToExit라는 케이스는 네이밍과는 별개로 존재할 이유가 없다고 생각해요. 위에서도 말씀드렸지만 게임 종료는 기능 아닐까요?

Error라고 생각하신 이유가 무엇인가요?

hyeffie replied:
이건 플레이어를 동등한 주체로 봐서 나왔던 네이밍이었던 것 같아요😂 구조 상 이름 자체는 exit 정도로 바꿔봐도 괜찮을 것 같아서 반영해볼게요.

에러를 사용한 이유는 위 코멘트에 대한 답변으로 대신할게요!

minsson commented on RockPaperScissors/RockPaperScissors/IO/IO.swift:
```diff
@@ -0,0 +1,26 @@
+
+import Foundation
+
+protocol PromptDisplayable {
+    func displayPrompt(_ prompt: String)
+}
+
+protocol InputGettable {
+    func getInput() throws -> String
+}
+
+protocol OuputDisplayble {
+    func displayOutput(_ output: String)
+}
+
+protocol HandGameErrorDisplayble {
+    func displayRPSError(_ error: HandGameError)
+}
+
+typealias IO = InputGettable & PromptDisplayable & OuputDisplayble & HandGameErrorDisplayble
```
딱 봤을 때는 무척 멋있어 보였는데, 프로토콜을 왜 이 정도로 세부적으로 나누셨는지는 잘 공감되지 않아요. 이 모든 게 Console의 역할이라고 생각해서요. 하나의 프로토콜로 했어도 충분했을 것 같아요.

또, 왜 protocol이 필요하다고 생각하셨나요? 앱을 만드는 상황까지 고려하신 건가요? 

두 분의 의견이 궁금해요.

hyeffie replied:
지금 구현을 보면 플레이어는 입력부터 출력까지의 기능을 모두 가진 입출력 객체를 가지고 있고, App은 에러를 출력하는 기능의 객체를 가지게 돼요.

실제로 지금 코드에서는 Console 클래스의 인스턴스 하나를 전역에서 공유하게 되지만(전역에서 같은 인스턴스를 공유하려는 의도는 없었는데 코드를 수정하는 과정에서 약간 실수가 있었습니다..) 사실 설계의 의도로 본다면 App의 에러 출력 속성에는 Console이 아닌, 다른 ErrorDisplayable 타입이 와도 돼요. 

만약 App에서 Console 타입을 사용한다면 실제로 사용하는 기능 외에 다른 기능에도 접근할 수 있을 거예요. 

저희의 의도는 사용하지 않을 다른 기능에 접근하지 않도록 하는 것이었고, 이렇게 객체가 사용하는 기능을 타입으로서 제한하도록 하는 패턴을 프로젝트 전반에 적용하고 싶었던 것 같아요.

minsson replied:
저는 이런 프로토콜의 사용이 과하다고 생각했어요. 하지만 이런 방식이 장기적인 방향에서는 좋은 접근이 될 수 있다고 생각해요. 캠프 과정에서 여러 프로젝트에 적용해보고, 연습해보면서 나중에 현업에서 어떻게 할지 정해보는 것도 좋을 것 같아요! 저는 이쪽 영역에서는 두 분이 저보다 잘하시는 것 같아서 과하다고 생각하는 것과는 별개로 두 분이 대단하시다고 생각했습니다! 😁

minsson commented on RockPaperScissors/RockPaperScissors/Player/MJBPlayable.swift:
```diff
@@ -0,0 +1,20 @@
+    func display(result: MJBResult)
+}
+
+protocol CallablePlayer {
+    var name: String { get }
+}
```
CallablePlayer라는 protocol을 만든 이유가 궁금해요. 코드의 가독성과 코드의 확장성 측면에서 어떤 고민을 통해 만든 프로토콜인가요?

hyeffie replied:
이건 플레이어가 게임에서 프로토콜 타입으로 사용된다는 가정 하에, 프롬프트에 현재 턴을 쥐고 있는 플레이어를 각 플레이어의 입력 프롬프트에 출력해야 하는 구조 때문에 필요한 요구사항이었어요.

```
[사용자 턴] 묵(1), 찌(2), 빠(3)! <종료 : 0> : 6
```

이런 식으로 말이죠…!

저희 구조에서는 플레이어가 Playable 이라는 타입으로 사용되고 있어서 Callable을 상속으로 강제하는 방식으로 구현했던 것 같아요.

안 그래도 PR을 올린 후였나.. 정작 name을 사용하는 쪽에서는 Callable이 아닌 Playable 로 접근하고 있는 걸 발견해서 수정본을 보관해두었는데 반영하면 좋을 것 같아요!

minsson commented on RockPaperScissors/RockPaperScissors/Player/ComputerPlayer.swift:
```diff
@@ -0,0 +1,28 @@
+    }
+}
+
+extension ComputerPlayer: CallablePlayer {
+    func getName() -> String {
+        return self.name
+    }
+}
```
이 코드는 사용되지 않는 코드로 보이네요. 사용하지 않는 코드는 제거하셔야 나중에 헷갈리지 않아요.

그런데 이 getName이라는 메서드는 ComputerPlayer, UserPlayer 모두 중복으로 정의 되어 있는데요. CallablePlayer라는 프로토콜이 존재하는 게 적절한지와는 별개로, 이왕 만들었으니 해당 프로토콜을 활용해 두 객체의 중복 코드를 제거하려면 어떻게 해야하는지 말씀해주실 수 있나요?

hyeffie replied:
이 부분은 말씀하신 것처럼 구현을 제거하지 않은 실수가 맞아요… 

별개로 중복 코드 제거에 대해 답변해보자면, 프로토콜의 기본 구현을 말씀하신 것 같은데 제대로 이해한 게 맞을까요?

```swift
protocol CallablePlayer {
    var name: String { get }
    
    func getName() -> String
}

extension CallablePlayer {
    func getName() -> String {
        return self.name
    }
}
```

물론 속성이나 getName 메소드 중 하나를 요구하는 방식이 더 깔끔하겠지만, 같은 구현을 적용하려면 위와 같은 방식이어야 할 것 같아요!

minsson replied:
넵, 프로토콜의 기본 구현을 말씀드리고 싶었어요. getName()이라는 메서드를 보자마자, protocol을 썼는데 왜 이 기능을 쓰지 않았을까 의아했어요. 물론 말씀하신 것처럼 name과 getName은 불필요한 중복으로 보이므로, 하나만 있으면 되겠네요.

minsson reviewed (changes requested):
안녕하세요, 노움, 에피!
두 분 모두 프로젝트 수행하시느라 고생하셨습니다!

코드와 질문을 보고 두 분이 얼마나 많은 고민을 하셨고, 얼마나 노력하셨는지 눈에 보여 코드리뷰를 하는 내내 많이 놀랐어요. 두 분의 열정을 보며, 저번 주와 마찬가지로 이번 주에도 많은 자극을 받았어요

이번 STEP에서는 Protocol의 활용에 많이 집중해주셨던 것 같아요. 저번에도 STEP에서도 느꼈지만, 구조에 대한 관심이 무척 많아보이고, 이 부분에서 다양한 시도를 하시며 성장하고 계신 것 같아요. 저도 이런 걸 공부하며 성장하려고 노력하고 있어서 두 분의 노력이 이해가 됩니다!

다만 저는 코드를 읽으며 좀 의아한 기분이 들었어요. `코드를 스윽 훑어봤을 때는 코드가 전반적으로 무척 깔끔하고, 추상화도 무척 잘 되어 있는 것 같아 보이는데, 왜 이렇게 잘 읽히지 않고, 이해하기가 어려울까?` 라는 생각을 했어요. 분명 클린한 코드 같은데, 가독성은 좋지 않게 느껴졌어요.

일단은 제 실력이 모자라서 그런 걸수도 있다는 생각을 했어요. 그러다 다른 한편으로는 프로토콜을 통한 추상화가 무척 과해서 그럴 수도 있겠다는 생각이 들었어요. 이 조그만 프로젝트에 9개의 프로토콜이 나오고, 각 프로토콜의 Requirement도 모두 1개예요. 

코드를 처음 읽는 제 입장에서는 매번 프로토콜의 의미를 파악해야 했고, 프로토콜의 선언부를 보며 프로토콜의 의도가 무엇인지 파악해야했고, 프로토콜을 채택한 타입이 왜 그 프로토콜을 채택해야 했는지, 어떤 기능이 있는지 등을 파악해야 했어요. 이 과정에서의 피로감이 느껴졌어요. 또, 대부분의 경우 프로토콜의 필요성도 잘 공감되지 않아 더 많은 피로감을 느꼈던 것 같아요.

이와 관련해 저도 좋은 코드를 고민하며 보게 된 영상이 하나 있었는데, 제게 도움이 됐던 내용이라 두 분에게도 공유 드려요. 한번 보시면 좋을 것 같아요. 어떤 생각이 드셨는지도 말씀해주시면 좋을 것 같아요.
[코딩애플 - 코딩 책 한 권만 읽으면 이렇게 됩니다](https://youtu.be/th7n1rmlO4I)

아래에는 두 분이 하셨던 질문에 대한 답변을 남겨요. 상세한 사항은 개별 코멘트를 참고해주세요.

두 분 고생하셨습니다!



## 질문에 대한 답변
### Player 가 IO를 가지고 있는 구조, 괜찮을까?
> 지금은 플레이어가 직접 자신의 인터페이스를 사용해 입력을 받고, 결과를 출력할 수 있도록 의도적으로 IO를 가지고 있도록 설계했는데요.

> 프로토콜을 사용해 결합도를 낮추고 있기는 하지만, 입출력을 소유하고 있는 구조에 대해서 스스로도 맞는 설계라고 확신하지 못하고 있습니다.

> 문제에 대해서 민쏜은 어떻게 생각하시는지 궁금합니다!

결합도를 말씀해주셨는데, 이 부분은 강한 결합, 느슨한 결합과는 무관한 부분으로 보여요.
객체의 역할에 대한 고려가 결합도에 대한 고려보다 더 중요하다고 생각해요. 그런 측면에서 좋은 구조로 보이지는 않아요. Player라는 객체의 역할이 모호해졌어요. 이런 역할은 Game 객체 또는 App 객체에서 해야하는 역할이라고 생각해요.

### class로 구현된 플레이어를 struct로 변경할 수 있을까?
> 현재 구현에서는 Playable 타입의 플레이어를 동일 비교하는 로직이 필요해서, RPSPlayable과 MJBPlayable을 클래스 전용 프로토콜로 선언하고 있는데요.
```swift
protocol RPSPlayable: AnyObject {
    func makeRPSGesture() throws -> RPSGesture
}
```
> 지난 리뷰를 참고하다보니 Player 객체들이 클래스로 선언되어야 할 이유가 없어서 구조체로 변경을 시도해보았습니다.

> 그런데 프로토콜 타입은 Equatable을 직접 구현할 수 없어서 결과적으로 구조체로 코드를 변경하는 데 실패했는데요.

> 혹시 플레이어를 동일 비교가 아닌 동등 비교할 수 있다고 볼 수 있을지, 그리고 현재 코드에서 클래스로 선언된 플레이어를 구조체로 변경할 수 있는 방법이 있는지 궁금합니다.

정확한 맥락이 쓰여있지 않아 잘 이해되지는 않지만, 정황상 `===`를 `==`로 바꾸는 과정에서 막히신 것 같아요. 아쉽지만 저도 이 부분에 대한 해결책이 마땅히 떠오르지 않아요.

`프로토콜 타입은 Equatable을 직접 구현할 수 없어서 결과적으로 구조체로 코드를 변경하는 데 실패했는데요.`라고 말씀하신 건 잘 이해되지 않아요. 결국 승자 판정 로직을 그대로 유지하면서 발생한 문제이므로, 승자 판정 로직을 조금만 바꿔도 해결하실 수 있었을 거예요.

가장 단순하게 생각해보면 enum 타입 객체를 두고, 상황에 맞게 승, 패, 무승부에 대한 case를 반환해 묵찌빠 게임의 진행여부를 분기처리해도 될 것 같아요.

아니면 지금의 승자판정 로직을 그대로 유지해야 할 이유가 있나요? 명확한 이유가 있다면, Player를 struct로 바꾸지 않아도 전혀 상관 없습니다.

두 분은 어떻게 생각하시나요?

### 플레이어의 생성 시점

> App은 가위바위보 게임과 묵찌빠 게임을 진행하기 위해서 게임을 할 수 있는(HandGamePlayable) 두 명의 플레이어와, 게임 진행 중 발생하는 에러의 메시지를 출력하기 위한(HandGameErrorDisplayble) 객체를 생성 시점에 전달받고 있는데요.

```swift
// main.swift
let console = Console()
let duo = HandGameDuo(UserPlayer(io: console, name: "에피"), ComputerPlayer())
HandGameApp(playerDuo: duo, errorDisplay: console).run()
```
> 그런데 객체와 메소드의 의미를 따져보았을 때, App 을 실행하기 전에 플레이어가 생성되는 것이 맞나? 앱을 실행했을 때 플레이어가 생성되어야 하는 것이 아닐까? 라는 의문이 들었습니다.

> 앱을 실행하는 시점에 플레이어를 생성하는 로직으로도 바꿔보았는데, 이렇게 되면 앱의 역할이 무엇인가? 이 로직들을 main에 나열하는 것과는 어떻게 다른가? 라는 또 다른 의문이 생기더라고요.

> 현재 App의 구현으로 보았을 때, 플레이어의 생성 시점은 어느 위치가 적절하다고 보시는지 민쏜의 의견이 궁금합니다.

코드리뷰를 하는 시점 기준에서는 PR 본문의 코드와 실제 최신 상태의 코드가 다르네요. 이 부분은 특별히 정답이 없어보여요. 다만 추상화된 App의 역할을 생각해봤을 때, 저는 App에서 Player를 만드는 게 자연스러워 보여요. 즉, 최신 상태의 코드가 적절해보여요.

이 질문과는 별개로 Console에 대한 게 좀 궁금한데, 이건 코드에 코멘트를 달아볼게요.


### 플레이어들이 묵찌빠를 할 수 있는지 확인하는 로직의 위치

질문의 길이가 길어 질문하신 내용 자체는 생략했습니다! 바로 답변 드릴게요.

`guard var mjbGame = MJBGame(rpsWinner: rpsWinner, rpsLoser: rpsLoser) else { return }`

위 코드에 관한 질문으로 보여요.

저는 묵찌빠 게임을 할지 말지는 App 객체가 결정하는 게 더 자연스럽고, 이해하기 쉽다고 생각해요. 저 개인적으로는 App에서 게임 전반의 흐름을 모두 관리하는 게 좋다고 생각하기 때문이에요.

현재 구조는 실패가능한 이니셜라이저를 굉장히 센스있게 사용한 느낌이 들어요. Swift 문법 공부를 여러모로 많이하신 것 같아요. 문법을 다양하게 활용해보는 건 무척 좋아요.

다만, 이 문맥에서 실패가능한 이니셜라이저를 쓰는 게 코드 가독성 측면에서 좋은 선택인지에 대한 의문이 들어요. 이 코드를 처음 보는 사람이 맥락 없이도 이해할 수 있는 코드일까요? 가위바위보와 묵찌빠의 정확한 관계를 모르는 사람의 경우 주석 없이는 이해하기 어려울 것 같아요. 또, 알고 있다고 해도 바로 파악하기 어려울 것 같아요.

저도 코드를 처음 볼 때 왜 실패가능한 이니셜라이저가 쓰인 건지 고민했어요. 제 실력이 모자라서인 걸수도 있지만요 😅

hyeffie commented:
### **Player 가 IO를 가지고 있는 구조, 괜찮을까?**

결합도 보다 객체의 역할이 중요하다는 말이 정말 중요한 것 같아요. 

사실 지난 리뷰에서 제안해주셨던 것처럼 `UserPlayer`, `PCPlayer`를 없애고 사용자인지 컴퓨터인지를 속성 만으로 표현하는 구조로도 구현을 해보았었는데요.

돌고 돌아 생각해보니 저는 처음에 이 게임이 한 명의 사용자만을 위한 게임이라는 걸 이해하지 못했던 것 같아요. 사용자와 컴퓨터는 동등한 주체라고 생각했고, 게임에 참여하는 플레이어는 그저 우리가 앱을 실행하며 사용하는 콘솔을 사용하는 것 뿐이라고 생각했어요.

그런 생각을 하고 나니까 자연스럽게 특정 플레이어 관점으로 결과를 생산하는 게임의 구조가 맞을까? 하면서 특정 사용자 관점에서 벗어난 결과를 리턴하는 게임을 만들고자 하는 생각이 들었고요.

플레이어를 만들고 보니 주체성을 갖게 된 플레이어에 과몰입 ? 하게 되면서 플레이어 스스로 (사용자처럼 출력이 가능한 상태라면) 결과를 출력해야 할 수 있으면 좋겠다고 생각하고, 이런 기능을 표현하는 프로토콜을 비롯해 입출력 속성(IO)까지 구현했었던 것 같아요.

정리하자면, 플레이어를 게임의 입출력에서 분리하려면 플레이어만의 입출력이 필요했고 그래서 속성으로 받게 되었어요. 의도는 그렇지만 다 구현하고 보니 인터페이스 타입이든, 객체이든 플레이어가 입출력을 직접 관장하는 것이 맞나라는 의문이 들었어요. 

또 여전히 종료나 에러에 대한 메시지를 출력하는 로직은 또 App이 가지고 있거든요. 플레이어 입출력으로 출력되지 않아요. 이 문제도 PR 을 보내기 전에 인지하고 있었는데, 플레이어가 이 케이스를 처리하도록 하는 해결책은 찾지 못했어요.

이렇게 풀어내고 나니 플레이어가 입출력을 직접 하게 됨으로써 구현이 요구사항으로부터 정말 많이 멀어졌다는 걸 느껴요. 사실 이 부분을 당장 해결하려면 설계가 많이 달라질 것 같아서, 사용자가 출력하는 방향으로 구현을 수정하는 건 더 많은 고민이 필요할 것 같습니다.

대신 플레이어를 굳이 분리하지 않고, 사용자 관점으로 결과를 리턴하던 이전의 설계에 대한 참조를 두고 갈게요. 

https://github.com/hyeffie/ios-rock-scissor-paper/tree/step2/1-prototype

### **class로 구현된 플레이어를 struct로 변경할 수 있을까?**

리뷰를 읽고 다시 한 번 struct 로 바꿔보려고 했는데, 프로토콜을 existential type으로 사용하는 한 동등 비교가 어렵더라고요. 이 부분이야 말로 제가 방법을 몰라서 바꾸지 못하는 건가 싶은데, 나중에 좀 더 공부해봐야 할 것 같아요. 🥺

사실 플레이어를 만든 의도라면 클래스로 구현한 의도가 아주 없다고 보기는 어렵거든요. 그래서 일단은 구현을 유지하기로 결정했어요.

### 마치며..?

민쏜의 코멘트를 읽고 전반적으로 코드에서 나타내려고 했던 부분들, 그리고 어쩌다 이런 설계를 하게 되었는지 연유에 대한 저희의 설명이 충분하지 않았다는 걸 많이 느꼈어요.

그래서 이런 상황에도 불구하고 최대한 의도를 읽어내고 이해해보려고 노력해주신 것 같아서 진심으로 감사한 마음이 들고요.

앞선 코멘트들에서도 얘기했지만, 사실 최종 설계 이전에 지난 스텝의 리뷰 사항들과, 두 번째 스텝의 새로운 요구사항을 만족하는 설계를 한 번 만들어 보기도 했어요.

그코 드가 완벽하다고 할 수는 없지만, 당시에 코드는 너무 단순하고 리뷰를 받기에는 고민이 충분하지 않다고 느꼈던 것 같아요. 그래서 사용자 플레이어 관점을 벗어난 구조 등의 다른 설계에 도전해 본 것도 사실이고, 코드적으로는 (인터페이스) 타입으로 기능을 표현하고 제한하는 코드를 작성해보려는 시도도 하면서 요구사항 이외의 것을 담아내려고 했던 것 같아요.

문제는 요구사항이라는 틀 안에서 이외의 것을 구현하려다 보니 결과적으로 읽기 어려운 코드가 되고, PR에서도 이런 의도들을 명확하게 드러내지 않았다는 거죠.

다음 프로젝트에서는 리뷰어가 의도를 읽을 수 있는, 설계 의도를 좀 더 선명하게 반영하는 PR 을 작성하는 연습을 해야 할 것 같아요. 또 아마.. 프로젝트가 복잡해지면서 요구사항 이외의 기능이나 구현을 추가할 여유가 없어질 것 같기도 하지만, 최대한 프로젝트와 각 스텝에서 해야 하는 경험에 초점을 두고 작업하도록 신경도 써보려고요.

다시 한 번 사려 깊은 리뷰와 코멘트들에 감사하다는 말씀드리고 싶어요! 2주간 너무 수고하셨습니다!

hyeffie committed:
fix #3: Console struct로 변경 및 구현 코드 분리

- Console struct로 변경
- 프로토콜 구현 코드 extension으로 분리

hyeffie committed:
rename #3: RPSGesture > RPSHand 로 이름 변경

- RPSGesture > RPSHand
- MJBGesture > MJBHand

hyeffie committed:
chore #3: import Foundation 제거

hyeffie committed:
chore #3: 게임 진행 가능 로직 App 레벨로 이동

hyeffie committed:
refactor #3: CallablePlayer 프로토콜 제거

- CallablePlayer 프로토콜 제거
- getName 구현 제거

hyeffie committed:
chore #3: App용 전역 console 사용하는 코드 제거

- App용 전역 console 사용하는 코드 제거
- 사용자 vs. 사용자 플레이용 코드 추가 (주석)

hyeffie committed:
feat #3: 게임 종료시 종료 메시지 출력 로직 수정

- HandGameErrorDisplayblePlayer 추가
- 게임 종료시 플레이어들이 자신의 IO에 종료 메시지 출력하도록 로직 수정 (App 에서 로직 이동)
- App의 errorDisplay 제거

hyeffie committed:
rename #3: 게임 종료 에러 이름 변경

hyeffie committed:
Merge branch '3-fix-step-2-review' into d_Gnome

hyeffie commented:
@minsson 리뷰 사항 반영했습니다!

중간에 App에서 에러를 display하는 로직에 대해서 고민하다가 IO 를 가지고 있는 플레이어의 원래 의도를 살려서 플레이어가 에러를 출력하도록 다시 구현해봤어요. 플레이어가 입출력을 가지고 있는 구조에 대해서는 저 스스로도 여전히 회의적이지만, 플레이어를 이런 의도로 분리했던 거구나. 라는 느낌으로 봐주시면 좋을 것 같아요.

아, 그리고 클로징에 위에 첨부해주셨던 영상에 관한 내용이 빠졌더라고요.

영상을 보면서 다 우리 코드에 해당 되는 내용이잖아 ... 😇 했던 것 같아요. 기능을 위해서 너무 많은 타입을 사용해서 코드를 분리하는데, 지금 구현에서는 타입을 재사용하지도 않아요. 심지어 어떤 의도를 가진 타입인지에 대한 설명이 주석에도, PR에도 없으니 개발 과정과 의도를 직접 읽어내느라 부담이 있으셨을 것 같아요 🥲

저도 전에 같은 요구사항에 대한 다른 사람의 코드를 읽으면서 이틀을 꼬박 썼던 적이 있어요. 재밌는게, 저도 처음에는 내 실력이 부족해서 좋은 구조의 코드를 읽기 힘든 걸까 생각했었어요. 결국 코드의 의도들은 알아냈지만, 그 코드는 거기에 머물렀지 만들었던 여러 인터페이스들을 활용하는 모습까지를 볼 기회는 없었어요.

재사용하지 않을 인터페이스, 타입, 함수를 꼭 만들어야 할까? 그건 영상의 주제처럼 다양한 의견이 있을 수 있는 문제라고 생각해요.  어떤 의미에서 영상에서 그런 주장을 했고, 민쏜이 이 영상을 공유해주셨을지 여러 고민을 거친 지금에서야 온전히 이해하게 된 것 같아요.

다만 주어진 요구사항에 추가로 게임을 바라보는 다른 관점으로도 설계를 해보고 싶은 욕심이 있었고, 그 내용을 민쏜에게 기회가 있을 때 리뷰 받아보고 싶었다는 정도로 이해해주시면 좋을 것 같습니다 ...🥹

그럼 리뷰 내용은 천천히 확인해주시고, 좋은 주말 보내세요! 수고하셨습니다 !

minsson commented:
안녕하세요, 에피, 노움!
2주 동안 고생 많으셨어요!! 

객체의 역할과 책임을 분리하는 건 객체지향에서 가장 기본이 되는 개념이자, 가장 어려운 개념이라는 생각이 들어요. 몇 년 후에 저희가 좀 더 성장하면 생각이 또 바뀔지도 모르겠지만요. 학습과정이 진행되는 동안 다양한(저보다 훨씬 더 잘하시는) 리뷰어분들이 함께 하실테니, 다양한 시도를 해보며 여러 피드백을 받아보셨으면 좋겠어요. 예를 들어 프로토콜을 많이 사용하신 것과 관련해, 저는 과하다고 했지만 그게 바람직하고, 더 좋은 방향이라고 말씀해주시는 분들도 분명히 계실 거예요. 

제가 참고용으로 드린 영상도 그 개발자 분의 개인적인 생각이니 영상에 달린 댓글로 반대 의견도 살펴보시면 좋을 것 같고요. 사람마다 생각하는 게 다르고, 맥락에 따라서도 말이 달라질 수 있으니 여러 의견을 둘러보며 자신의 기준을 찾아나가시면 될 것 같습니다! 

이 여정을 시작하기도 쉽지 않고, 하면 할 수록 머리가 아파져서 지속하는 것도 쉽지 않은데요. 두 분은 이미 그 여정이 한참 진행 중인 것 같아서 두 분 모두 대단하다는 말씀을 드리고 싶어요. 두 분 덕분에 구조에 대해 더 많이 고민할 수 있었고, 두 분이 주신 자극 덕분에 업무에서 새로운 시도를 해보기도 했어요. 이래서 코드리뷰를 하며 함께 성장한다는 말이 있다는 생각도 들었고요.

지금의 열정을 학습 과정 내내, 그리고 그 이후로도 유지할 수 있기를 바랄게요!
2주 동안 정말 고생하셨고, 다음 프로젝트도 화이팅이에요! 🚀
